adobe/react-spectrum
# [useButton & useToggleButton]: Add dynamic ref type depending on elementType props

Closes #1172

Hello Adobe team 👋 
This is an implementation proposal to the types modification request in response to the following issue: #1172

I modified four files :
- `useButton` and `useToggleButton` hooks on which I added functions overloading (thanks to @devongovett advices) to dynamically modify the generic type passed to the returned `ButtonAria` depending on `elementType` props. I also modified the ref object generic type `HTMLElement` to match with the `elementType` one.
- button types in `@react-types/button`: changed `elementType` type from string to `React.ElementType` which may enforce typing to be sure `buttonProps` ref and `elementType` are corresponding to the HTML DOM element used.
- Button spectrum element for a small generic type update

## ✅ Pull Request Checklist:


--> Did not added testable code as my modifications are mainly related to types. Please tell me, otherwise, if it may need some.

--> I assumed the props are automatically generated based on the component's props. If no, could you tell me where to update them on `docs:@react-aria/button` ? Also, I think restriction HTML types who would make sense to be used as a button should be notified on the documentation

## 📝 Test Instructions:

1. Create a button component :

```typescript
export function Button(props: AriaButtonProps) {
  const ref = React.useRef<HTMLDivElement | null>(null);
  const {buttonProps} = useButton(
    {
      ...props,
      elementType: 'div'
    },
    ref
  );
  return (
    <div {...buttonProps} ref={ref}>
      {props.children}
    </div>
  );
}
```

2. Change `useRef` type, `elementType` and HTML node to whatever you want
3. `buttonProps` type should be good

## 🧢 Your Project:

Personal project -> would like to use react-aria with custom components using Framer Motion which may be have different DOM element.
